openedx-unsupported/configuration
# Create Analytics Seed Job

Replaces PR #2830.

Creates and builds the [AnalyticsSeedJob](https://github.com/open-craft/edx-jenkins-job-dsl/blob/jill/emr-analytics-pipeline-task/jobs/analytics-edx-jenkins.edx.org/createJobs.groovy) (see [PR-15](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/15/files)) to create analytics task jobs in Jenkins for each configured analytics task.

Uses [analytics task and parameters configuration from the ansible vars](https://github.com/open-craft/configuration/blob/jill/jenkins-analytics-dsls/playbooks/roles/jenkins_analytics/defaults/main.yml#L28) to initialise the seed job git repo details and default `EXTRA_VARS` parameter value.

**JIRA tickets**: Implements OC-1370, OLIVE-20

**Discussions**: Ongoing discussions with @e0d and @mulby , @jbzdak , @mtyaka

**Screenshots**:
- Jenkins home page after running ansible:
  ![jenkins home](https://cloud.githubusercontent.com/assets/7556571/14147632/d76fd972-f6e3-11e5-90f6-440b6a9f364a.png)
- AnalyticsSeedJob build parameters (initial build and defaults set by Ansible):
  ![analyticsseedjob build parameters](https://cloud.githubusercontent.com/assets/7556571/14148324/91d4eb70-f6e6-11e5-90e4-0f51044e45e4.png)
- AnswerDistributionWorkflow  analytics task shell command:
  ![answerdistributionworkflow task shell command](https://cloud.githubusercontent.com/assets/7556571/14097381/b19d1818-f5b7-11e5-9d09-b43bb7b0b6bf.png)
- AnswerDistribution analytics task build parameters:
  ![analytics task build parameters1](https://cloud.githubusercontent.com/assets/7556571/14147661/ff4d9826-f6e3-11e5-9c60-37e45826274e.png)
  ![analytics task build parameters2](https://cloud.githubusercontent.com/assets/7556571/14147671/06d3194a-f6e4-11e5-8332-4095174b4f58.png)
  ![analytics task build parameters3](https://cloud.githubusercontent.com/assets/7556571/14147684/0a37248c-f6e4-11e5-98df-76e0049c6c1b.png)
  ![analytics task build parameters4](https://cloud.githubusercontent.com/assets/7556571/14147688/0e065920-f6e4-11e5-8ced-a4c75db32a8e.png)
- CourseActivityWeekly shell command:
  ![courseactivityweekly shell command](https://cloud.githubusercontent.com/assets/7556571/14147990/46cf47f2-f6e5-11e5-84ea-7112c4ae30d7.png)

**Sandbox URL:** http://52.20.136.133:8080/

Testing instructions:

To set up a jenkins_analytics VM, see PR #2810 for instructions, and note changes:
- Checkout my branch, not Jacek's: open-craft:jill/olive-22-jenkins-analytics-dsls
- You don't need access to [edx-ops/edx-jenkins-job-dsl](https://github.com/edx-ops/edx-jenkins-job-dsl/), but you do need access to [open-craft/edx-jenkins-job-dsl](https://github.com/open-craft/edx-jenkins-job-dsl.git)
- See `Settings` below for example settings to use in your `ENV[VAGRANT_JENKINS_LOCAL_VARS_FILE]`
- Once you login to Jenkins, instead of `stub-analytics-jobs`, you'll see `AnalyticsSeedJob`

Or to use the sandbox:
1. Login to Jenkins http://52.20.136.133:8080/ .  Contact me for password.
2. Click on the AnalyticsSeedJob.
3. Click on Configure to review configuration, and see the seed configuration parameter EXTRA_VARS
4. Click on Build with Parameters to build.  Build will fetch `open-craft/edx-jenkins-job-dsl` and run the DSLs, which creates or updates the analytics jobs.
5. Return to Jenkins home, and click on a generated analytics job. 
6. Building the analytics job will call `run_automated_task.sh` (See notes below).
7. View Console on each Job to see the output.

**Sandbox Setup:**

@mtyaka set up the sandbox for [OC-1384](https://tasks.opencraft.com/browse/OC-1384):
1. [initial deployment](https://gist.github.com/mtyaka/1ea3fd145863ee6d3173)
2. Seed task and DSL deployment: 
   
   ```
   cd /home/jenkins/configuration
   git checkout -b jill/olive-22-jenkins-analytics-dsls origin/jill/olive-22-jenkins-analytics-dsls
   source /home/jenkins/workspace/venvs/configuration/bin/activate
   pip install -r requirements.txt
   ansible-playbook --connection local -i 'localhost,' --extra-vars=@../ansible-jenkins-credentials.yml playbooks/analytics-jenkins.yml
   ```
3. Copy the SSH keys created during the initial deployment to the new Jenkins home dir:
   
   ```
   sudo -u jenkins cp -pi /home/jenkins/.ssh/* /edx/var/jenkins/.ssh/
   ```

**Author notes and concerns**:
1. Analytics tasks do not run, but EMR provisioning/terminating does.

**Reviewers**
- [] Code: (@jbzdak, @e0d )
- [] Product: (TBD)
- [] Platform: (TBD)
- [] UX: (TBD)
- [] TNL: (TBD)
- [] a11y: (TBD)
- [] doc/strings: (TBD)

**Settings**

See [configuration playbooks/roles/jenkins_analytics/README.md](https://github.com/open-craft/configuration/blob/jill/olive-22-jenkins-analytics-dsls/playbooks/roles/jenkins_analytics/) for details.

Create a YML file outside your repo, and `export VAGRANT_JENKINS_LOCAL_VARS_FILE=/path/to/file.yml` to that location.  Follow these steps to populate the `VAGRANT_JENKINS_LOCAL_VARS_FILE`.
1. Choose a password for the jenkins system user for `JENKINS_ANALYTICS_USER_PASSWORD_PLAIN`
2. Get access to the [jobs DSL repo](https://github.com/open-craft/edx-jenkins-job-dsl/tree/jill/dsl-closures) used in [PR #15](https://github.com/edx-ops/edx-jenkins-job-dsl/pull/15)`, point`JENKINS_ANALYTICS_GITHUB_KEY` to that user's key file.
3. Set up ssh keys for the jenkins user to access the AWS/EMR cluster, as this step is not handled by this work.

See [ansible-jenkins-credentials.yml.zip](https://github.com/edx/configuration/files/197452/ansible-jenkins-credentials.yml.zip) for the (sanitised) configuration used for the analytics sandbox.
1. Run `vagrant up`
